File: src/main/java/jp/pigumer/ClasspathScanner.java
package jp.pigumer;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;
import java.util.Optional;
import java.util.function.BiFunction;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import java.util.stream.Stream;
import java.util.zip.ZipEntry;

public class ClasspathScanner {

    private static final String JAR_URL_SEPARATOR = "!/";

    URL getUrl(URL url) throws MalformedURLException {
        String file = url.getFile();
        int index = file.indexOf(JAR_URL_SEPARATOR);
        if (index != -1) {
            String jarFile = file.substring(0, index);
            try {
                return new URL(jarFile);
            } catch (MalformedURLException e) {
                if (jarFile.startsWith("/")) {
                    return new URL("file:" + jarFile);
                }
                return new URL("file:/" + jarFile);
            }
        }
        return url;
    }

    Optional<JarFile> getJarFile(URL url) {
        try {
            URI uri = getUrl(url).toURI();
            return Optional.of(new JarFile(uri.getSchemeSpecificPart()));
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    public <R> R scan(ClassLoader classLoader, String path, BiFunction<String, Stream<ZipEntry>, R> mapper) throws IOException {
        Enumeration<URL> urls = classLoader.getResources(path);
        List<JarFile> list = new ArrayList<>();
        while (urls.hasMoreElements()) {
            getJarFile(urls.nextElement()).ifPresent(jarFile -> list.add(jarFile));
        }
        return mapper.apply(path, list.stream().flatMap(jarFile -> {
            Stream<JarEntry> entry = jarFile.stream();
            return entry.filter(item -> item.getName().startsWith(path));
        }));
    }

    public <R> R scan(Package rootPackage, BiFunction<String, Stream<ZipEntry>, R> mapper) throws IOException {
        String rootName = rootPackage.getName();
        String path = rootName.replaceAll("\\.", "/");
        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
        return scan(classLoader, path, mapper);
    }
}

File: src/main/scala/jp/pigumer/ClasspathScanner.scala
package jp.pigumer

import java.io.File
import java.net.{MalformedURLException, URL}
import java.util.jar.JarFile

import collection.JavaConversions._

class ClasspathScanner {

  private def getUrl(url: URL): URL = {
    val file: String = url.getFile
    val index = file.indexOf("!/")
    if (index != -1) {
      val jarFile = file.substring(0, index)
      try {
        new URL(jarFile)
      } catch {
        case e: MalformedURLException ⇒
          if (jarFile.startsWith("/")) {
            new URL("file:" + jarFile)
          } else {
            new URL("file:/" + jarFile)
          }
      }
    } else {
      url
    }
  }

  private def getJarFile(url: URL): JarFile = {
    val uri = getUrl(url).toURI
    new JarFile(uri.getSchemeSpecificPart)
  }

  private def files(f: File, path: String): Seq[String] = {
    if (f.isFile) {
      List(path)
    }
    files(f, path + "/" + f.getName)
  }

  private def listFiles(url: URL, path: String): Seq[String] = {
    val root = new File(url.getFile)
    files(root, path)
  }

  private def listEntries(jarFile: JarFile, path: String): Seq[String] = {
    val entries = jarFile.entries()
    val filtered = entries.filter(e ⇒ !e.isDirectory && e.getName.startsWith(path))
    val list = filtered.map {
      entry ⇒ entry.getName
    }
    list.toList
  }

  def scan(classLoader: ClassLoader, path: String): java.util.List[String] = {
    val convertedPath = path.replaceAll("\\.", "/")
    val urls = classLoader.getResources(convertedPath)
    val seq = urls.flatMap {
      url ⇒ {
        val resourceUrl = getUrl(url)
        val resourceFile = resourceUrl.getFile
        if (resourceFile.endsWith(".jar")) {
          listEntries(getJarFile(resourceUrl), convertedPath + "/")
        } else {
          listFiles(resourceUrl, convertedPath)
        }
      }
    }
    seq.toList
  }
}